.pfnci/schema.yaml

```diff
@@ -0,0 +1,342 @@
+###############################################################################
+# System
+###############################################################################
+# Target system to test on.
+system:
+  "linux":
+  "windows":
+
+###############################################################################
+# OS
+###############################################################################
+# Name and version of the operating system to test on.
+os:
+  "ubuntu:18.04":
+    system: "linux"
+  "ubuntu:20.04":
+    system: "linux"
+  "centos:7":
+    system: "linux"
+  "centos:8":
+    system: "linux"
+  "ws:2016":
+    system: "windows"
+
+###############################################################################
+# CUDA & ROCm
+###############################################################################
+# The platform to test with.
+# `cuda` and `rocm` are mutually exclusive; when one is set, the other must be null.
+# `full_version` is the version used in base Docker images.
+# When adding a new CUDA version, you may also need to add the version to the
+# list of supported CUDA versions in the optional libraries.
+cuda:
+  null:
+    full_version: null
+  "10.0":
+    full_version: "10.0"
+  "10.1":
+    full_version: "10.1"
+  "10.2":
+    full_version: "10.2"
+  "11.0":
+    full_version: "11.0.3"
+  "11.1":
+    full_version: "11.1.1"
+  "11.2":
+    full_version: "11.2.1"
+  "11.3":
+    full_version: "11.3.1"
+  "11.4":
+    full_version: "11.4.0"
+rocm:
+  null:
+    full_version: null
+    packages: []
+  "4.0":
+    full_version: "4.0.1"
+    packages: ['rocm-dev', 'hipblas', 'hipsparse', 'hipcub',
+               'rocsparse', 'rocrand', 'rocthrust', 'rocsolver', 'rocfft',
+               'rocprim', 'rccl']
+  "4.2":
+    full_version: "4.2"
+    packages: ['rocm-dev', 'hipblas', 'hipfft', 'hipsparse', 'hipcub',
+               'rocsparse', 'rocrand', 'rocthrust', 'rocsolver', 'rocfft',
+               'rocprim', 'rccl']
+
+###############################################################################
+# CUDA Optional Libraries
+###############################################################################
+# The optional libraries to install: NCCL, cuTENSOR, cuSPARSELt and cuDNN.
+# `spec` must be a version specifier accepted by apt and yum.
+# `cuda` lists the supported CUDA version.
+# `cuda.alias` (cuDNN only) is an alias name for package name suffix.
+nccl:
+  # https://docs.nvidia.com/deeplearning/nccl/archives/
+  # https://developer.nvidia.com/nccl/nccl-download
+  null:
+    spec: null
```

Comment:
Is this needed? seems that all libraries define it by default, but I didn't see any explicit check against it in the code.

Reply:
Good point, `spec: null` is not needed actually. Will remove.

Reply:
I tried to remove and recalled that this is as expected 😅 
In YAML, 

```yaml
key:
  subkey: 1
key2:
```

is equivalent to

```py
{'key1': {'subkey': 1}, 'key2': None}
```

so I need to add many null(None)-checking to `generate.py` when iterating the dict above. Defining subkeys as a dict:

```yaml
key:
  subkey: 1
key2:
  subkey: null
```

makes a life lot easier, as it will be loaded as:

```py
{'key1': {'subkey': 1}, 'key2': {'subkey': None}}
```

Reply:
understood! thanks for the explanation!